[chore] added style for file links with no permission

Kharonus opened the pull request.

Kharonus committed:
[chore] added style for file links with no permission

HDinger commented on frontend/src/global_styles/content/work_packages/tabs/_files.sass:
```diff
@@ -71,6 +71,20 @@
+        @include spot-body-small()
+
+        padding-right: $spot-spacing-0-25
+        color: #1A67A3
```
Is it by chance that this is the current primary color? So in other words: Shall this always be exactly this hexcode or rather the primary-colour? If the later, you can use `color: var(--primary-color)` instead

Kharonus replied:
you are probably right, this is text displayed in the primary color. changed.

HDinger commented on frontend/src/app/shared/components/file-links/file-link-list/file-link-list-item.component.ts:
```diff
@@ -106,4 +111,24 @@ export class FileLinkListItemComponent implements OnInit, AfterViewInit {
+        break;
+      case 'urn:openproject-org:api:v3:file-links:permission:NotAllowed':
+        this.showFloatingActions = false;
+        this.text.floatingText = this.i18n.t('js.label_file_link_no_permission');
```
Instead of adding the text strings here, please add them at the top (where the text is defined). Here you can that write something like `this.floatingText = this.text.foo`. You'd then need a public variable for that. However, it is more clear to have all used I18n strings in one place and not distributed all over the code.

Kharonus replied:
fixed

HDinger commented on frontend/src/app/shared/components/file-links/file-link-list/file-link-list-item.component.ts:
```diff
@@ -106,4 +111,24 @@ export class FileLinkListItemComponent implements OnInit, AfterViewInit {
+        this.showFloatingActions = true;
+        this.text.floatingText = '';
+        break;
+      case 'urn:openproject-org:api:v3:file-links:permission:NotAllowed':
```
We have a capabilities API with which you can fetch permissions for a user. Is there a reason you did not use that one here?

Kharonus replied:
Those are permissions fetched from nextcloud. we cannot deliver those via the capabilities API.

HDinger replied:
That's too bad :/ In that case I would either extract them to global constants so that you only have to manage one place. Or you create a service that does handles these special permissions.

Kharonus replied:
extracted to constants

HDinger commented on frontend/src/app/shared/components/file-links/file-link-list/file-link-list.html:
```diff
@@ -19,6 +19,7 @@
     <li
       *ngFor="let fileLink of fileLinks$ | async; let i = index;"
       class="spot-list--item op-files-tab--file-list-item"
+      [ngClass]="{ 'disabled': fileLink._links.permission.href !== 'urn:openproject-org:api:v3:file-links:permission:View' }"
```
Same as above. The capabilities API might be better. If you can't use it, please extract the `urn:openproject-org...` into constants at least so that we don't have to maintain multiple strings at different places.

HDinger commented on frontend/src/app/shared/components/file-links/file-link-list/file-link-list-item.component.ts:
```diff
@@ -106,4 +111,24 @@ export class FileLinkListItemComponent implements OnInit, AfterViewInit {
+
+  private derivePermissionState():void {
+    switch (this.fileLink._links.permission.href) {
+      case 'urn:openproject-org:api:v3:file-links:permission:View':
```
Is this string the same for every instance? So even for self-hosted, custom-domain instances it will start with `urn:openproject-org`?

Kharonus replied:
those constant URNs are static for the whole OP product. What I am currently missing, is some kind of central lookup in the frontend. I feel like it is very cumbersome, if we just put those into our components or templates.

Kharonus committed:
Merge branch 'dev' into chore/file-link-not-viewable-style

HDinger commented on config/locales/js-en.yml:
```diff
@@ -399,6 +399,8 @@ en:
     label_expand_project_menu: "Expand project menu"
     label_export: "Export"
     label_export_preparing: "The export is being prepared and will be downloaded shortly."
+    label_file_link_no_permission: "No permission to see this file."
```
If these are user shown error messages, I'd prefer a correct sentence like "You have no permission to see this file". Same for the other string.

Kharonus replied:
This was the message defined by this figma file: https://www.figma.com/file/gtLQfPe09X7XugAH8L7dTy/Nextcloud-Integration?node-id=1062%3A33627

I will contact Marc to discuss those texts.

HDinger replied:
👍

Kharonus replied:
permission text is fixed, sadly marc is not yet sure about how to proceed with the error texts. Which is why a change is not included yet.

HDinger reviewed (changes requested):
Even as a user with role "Reader", I still the the user hover action and none of the error messages. When clicking the user icon, I get an error. Which permission do I need (or not need) to see the effect?

**Edit** After the merge commit, I don't see the user but an empty shadow overlay:
<img width="518" alt="Bildschirmfoto 2022-06-27 um 13 24 02" src="https://user-images.githubusercontent.com/7457313/175930720-7f000995-f859-4208-b15f-55343700c3d7.png">

Kharonus committed:
[chore] extracted URN constants to separate file

- fixed localized texts
- used primary color for floating texts
- small refactoring in file links

Kharonus committed:
[chore] removec floating error text

Kharonus commented:
After talking to product team:
- floating error text is removed
- instead of floating error on every file link, file link list of a storage MUST NOT be rendered, if at least one file link has an error state (logically it will be always EVERY or NO file link with errors, but to get frontend behave consistently we hide list already on a single error)
- this behaviour is not implemented here yet, it will be implemented later, if the login from storage information box is merged. otherwise a bunch of merge conflicts is expected.

HDinger reviewed (changes requested):
Code-wise it looks good. I just have some styling related issues:

- [x] The transparency affects not only the file name but also the error text. In figma it is specified, that the text should have normal opacity.
<img width="400" alt="Bildschirmfoto 2022-06-28 um 12 55 01" src="https://user-images.githubusercontent.com/7457313/176162526-ae40fa16-9c22-4ca8-8865-80cc5ad2cae0.png">

- [x] When the error text is longer, it can span multiple lines overlapping the one below by half. Either you truncate the text, or you make sure, that both lines fit into the height of one file. 
<img width="507" alt="Bildschirmfoto 2022-06-28 um 13 01 45" src="https://user-images.githubusercontent.com/7457313/176163631-3e1b568e-66bf-4957-82dd-f5e8428f85dd.png">


- [x] Nothing prevents me from clicking on the file name and getting redirected to Nextcloud. Should that be possible, if I don't have the permission anyway to see the file?
- [x] From a security perspective, one could argue that displaying the name of the file already provides some Information which the user is actually not allowed to see.

Kharonus commented:
Issues commented:
- the transparency effect is the same for actions or text. Not sure, why in figma it isn't visible. I'll recap it once more with marc, but I'm pretty sure, we can leave it as is. (Edit: your concern is correct, I misunderstood the intention)
- truncating this text removes the whole purpose of a floating info text. I'll ask marc. But I think, we need to make sure, that the localized texts do not get any longer. From technical perspective we can of course prevent line breaks.
- clicking the file is still possible, that is intended. nextcloud itself has no "not allowed" page, but instead will render a not found toast, but in any case, the link itself is still a valid one. We can think about blocking it, but I would rather put this into a separate feature requests (As far as I understand it, figma still shows an underline suggesting a clickable link)
- this is a known issue and was well discussed in the nextcloud team. The information exposure is intended, to facilitate communication in teams.

Kharonus committed:
[chore] fixed style flaws

Kharonus commented:
Additional information:
- long floating texts get truncated, though it should considered a localization flaw, if someone inserts a long text here.
- gradient is wanted for both floating text and actions
- clicking is wanted on not viewable file links
- floating text is no longer effected by the opacity reduction

HDinger reviewed (commented):
> long floating texts get truncated, though it should considered a localization flaw, if someone inserts a long text here.

The text is not really truncated but only cut off. However, this would be fine for me since it does not break the design any more.
<img width="493" alt="Bildschirmfoto 2022-06-28 um 14 10 51" src="https://user-images.githubusercontent.com/7457313/176175928-3f88ef3e-dced-450d-b792-32f44f44df8b.png">
